### ISSUE TITLE
Get rid of overload keyword

### ISSUE THREAD

**Issue by jckarter** [jckarter/clay#410]
The `overload` keyword is a wart, and it'd be nice if it weren't necessary. The best proposal I've heard for doing so is to change definition behavior as follows:
- Have `foo() { }` define a new standalone function `foo` in the current module if one doesn't currently exist, or overload a symbol or type constructor `foo` if it does. If `foo` is already defined as something other than a `define` or type, raise an error.
- Require qualified names to overload symbols from other modules, e.g. `core.operators.(+)(a:Foo, b:Foo) = a.foo + b.foo;`

**Comment by ghost** [comment]
A common occurrence in the lib is to define symbols in a base module, import and then overload, particularly since the recent import circularity issue was addressed. So this suggestion for `overload` keyword removal may lead to just as much typing of the qualified name as one would currently type `overload`. However this would not be much of an issue with import aliases:

```
import core.operators as cops;
cops.(+)(a, b) = a + b;
```

or

```
alias cops = core.operators;
cops.(+)(a, b) = a + b;
```

Either way I think it is more informative to use qualified names than `overload` everywhere.

**Comment by jckarter** [comment]
Yeah, you should be able to use a module alias too, as long as there's some qualification indicating that the definition is referencing an imported symbol rather than a new local symbol.

**Comment by jckarter** [comment]
The `define` keyword might more appropriately be called `generic` or `declare` now, since it really declares a generic symbol rather than providing any definition. So instead of this:

```
import widgets.(gizmo);

foo(x, y) { ... }

define bar(x, y);
overload bar(x:Int, y:Int) { ... }
overload bar(x:Float, y:Float) { ... }

overload gizmo(x:Int) { ... }
```

You'd have:

```
import widgets;

foo(x, y) { ... }

generic bar(x, y);
bar(x:Int, y:Int) { ... }
bar(x:Float, y:Float) { ... }

widgets.gizmo(x:Int) { ... }
```

**Comment by ghost** [comment]
Or even `template bar(x, y);` . . .

**Comment by stepancheg** [comment]
@jckarter module prefixes for overloads don't look good:
- You do not import `widgets.*`, so you have to prefix with module name all calls  from current module to functions in `widgets` module
- Overloads are noisy for lengthy module names
- It is not clear, whether

```
import printer;

printer.printTo(stream, i: Int) { .. }
```

is allowed (`printTo` is declared in `printer.protocols`).

Current explicit `overload` is OK to me, especially because it is explicit.

If you are going to rename `define`, my vote is for `declare`.

**Comment by jckarter** [comment]
`printer.printTo` should also work; my main intention with requiring qualification is to make it clear when an imported generic is being overloaded, not to require that the overloaded symbol be referenced only by its exact name. Retaining `overload` for this purpose is a possibility, but it seems redundant for local definitions now that `define` is required. After `define foo;`, `foo() {}` couldn't do anything but overload `foo`.

**Comment by stepancheg** [comment]
@jckarter `foo() {}` can't do anything but overload `foo`, when I encounter `foo() {}` without preceding `define foo` I'm not sure whether `foo` is one of overloads or standalone single implementation: I have to search for `define foo` above in the current file.

On the other hand, it is not important whether local or imported function is overloaded (especially if function is declared in internal submodule of current module, like `io.streams` overloads functions from `io.streams.protocols`), so manadatory prefix for imported functions seems to be redundant to me.

**Comment by jckarter** [comment]
What if standalone functions also required a keyword? That way every definition requires a keyword, and you'll have to think about the keyword rather than being tempted to elide it:

```
function foo() {} // standalone
declare bar(); // declare generic
overload bar() {} // define method
```

**Comment by galchinsky** [comment]
@stepancheg, with overload keyword you are also not sure whether it is standalone or not. You only know that there is a define somewhere. 95% it is nearby. Or somewhere you don't know. The number of functions is greater than zero. That is all.
So it's better a final function to be declared as a final, because it is only when you are sure.

```
import m.(++++);
import othermodule; 
fn foo(){} //i.e. final
(++++)(a, b){}
othermodule.qaz(){};
declare bar();
bar(){}
```

I don't like that 'forceinline overload...' in the whole library. It is a noise like 'public final static' in Java. I'd like to write forceinline{} and some other blocks like extern "C"{} and get rid of overload keyword.

```
forceinline {
    declare BoolAssoc?(..x) : Bool;
    declare LeftAssoc?(..x) : Bool;
    declare RightAssoc?(..x) : Bool;

    BoolAssoc?(..x) : Bool = false;
    LeftAssoc?(..x) : Bool = false;
    RightAssoc?(..x) : Bool = false;

    [OP] BoolAssoc?(#OP) : Bool = integerEquals?(operatorAssociativity(OP),BOOLEAN);
    [OP1,OP2] BoolAssoc?(#OP1,#OP2) : Bool = BoolAssoc?(OP1) and BoolAssoc?(OP2);

    [OP] LeftAssoc?(#OP) : Bool = integerEquals?(operatorAssociativity(OP),LEFT);
    [OP1,OP2] LeftAssoc?(#OP1,#OP2) : Bool = LeftAssoc?(OP1) and LeftAssoc?(OP2);

    [OP] RightAssoc?(#OP) : Bool = integerEquals?(operatorAssociativity(OP),RIGHT);
    [OP1,OP2] RightAssoc?(#OP1,#OP2) : Bool = RightAssoc?(OP1) and RightAssoc?(OP2);
}
```

**Comment by jckarter** [comment]
I don't think the preponderance of `forceinline` in the core is a huge issue, since in user code you shouldn't be using `forceinline`. I do agree however that attribute blocks would be very useful, not only for attributes but for pattern variables as well.

**Comment by stepancheg** [comment]
@jckarter 

About `function` keyword.

Since `overload` is also a function, `function` keyword for non-overload is not appropriate. Something like `standalone` would be better, however I think it is even better without a keyword, as it is now, because most functions in application code (not library) are standalone functions.

@galchinsky 

I don't think attribute repetition (or pattern variable repetition) is an issue. But, the necessity to scroll up to find which are current attributes will be a problem. One more language complication is also a problem.

In summary, I'm conservative about this part of the language. It is almost perfect. Let's proceed slowly.

**Comment by galchinsky** [comment]
@stepancheg
You say like `public:` and `private:` in class definition C++ is a problem, and public, public, public, public in Java is not. Scrolling is good. If everybody thought that user must have maximum info in a screen,  we would still have to use switches instead of any kind of polymorphism. A better answer here is some-kind-of-lint+lightweight syntax.
But I agree with your summary. This sugar will take a time and do little.

**Comment by jckarter** [comment]
I think the status quo can still be improved here; although I appreciate the explicitness of `overload` over the ad-hoc rules other languages adopt to hide it, empirically, it's a common stumbling block for programmers new to Clay, and I personally to this day reflexively leave it out where it's required. I think either eliminating the keyword or requiring keywords for all forms would be a bit better.

**Comment by stepancheg** [comment]
BTW, sometimes I think that some (or all) compiler errors should contain short link to a web page containing explanation of error, small code example producing similar error, and a link to specification. This would help newcomers start programming without reading the specification.

**Comment by jckarter** [comment]
That would be a nice cleanup; also, instead of `error` being called directly at every error site with custom-built strings,  error messages should be collected into a separate module as functions. It would then be easy to review the set of errors and add references to the wiki or some other external resource.

**Comment by jckarter** [comment]
Closing this for lack of a consensus on action.

**Comment by ghost** [comment]
Another possibility here is to allow overload groups, as per issue #423.

**Comment by ghost** [comment]
joe, i'd like to see this opened again. possibly, i tend to push a _semantic vs. pragmatic view debate_ (linguistically). though i mainly use flat, pragmatic, blue-print-style low-level languages, i like the abstraction powers of scheme, for example. scheme fails to provide a pleasing grammar though. clay really convinced me that it is out to target this issue. possibly i'm wrong with that. but when i look at clay modules, they consist of simple snippets in a logical order, much like i'd evolve and lay out an idea or concept on paper. clay script also provides a beautiul _form_ for this. but i'd like to see it evolve even more into direction of a protocol-style. consequently, the redundant overload keyword, which not only represents a weak technical term but also looks like an artefact from generic stoneage, should be dropped for what you proposed above. same with define/declare: though _generic_ is a bit too generic, i like it far more. _symbol_ is an alternative. i'd like clay to use terms like that (we already have _when_, but _given_ was also a good idea.)  _which_, instead, was a very bad idea because of the weird semantics.

given the above, i also don't see the point in not wanting to scroll in the editor window. because clay code is written a way that you will need to read from top. i rather care about a good coding style that gives a hint about where to search for what. possibly grouping is a way out. but there is already some semantic guideline in clay code by how it evolves. at least i see it so.
